xtm/xtm_scsp_queue.h

```diff
@@ -0,0 +1,184 @@
+#pragma once
+/*
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/**
+ * Lock free single consumer, single producer queue,
+ * based on ring buffer.
+ * Template parameter can be of any type for which the
+ * copying operation is carried out quickly and correctly.
+ */
+template<class T>
+struct xtm_scsp_queue {
+	/** Next position to be written */
+	unsigned write;
```

Comment:
The members should be private, since they can't be safely accessed without the methods of the class. The same for the iterator. But I don't know if we really want to go full C++ insanity here.

Reply:
Fixed. About C++:
Originally i wanted to make a void * queue and get / put functions only. I believe that even taking into account the current implementation of httpng, it is not difficult to remake the queue into a void * and pass the function in the message as well. However, the necessity to add xtm_fun_dispatch/invoke to API gave me the choice to create a rather ugly queue of `union xtm_msg` or or to use templates. I think templates is better, but if you dont think so i can create queue of `union xtm_msg`.

Reply:
With httpng they need to pass a function and an argument, which is 16 bytes. They won't fit into one pointer. It means they would need to allocate 16 messages on the heap. I suspect this is the main reason they didn't want to use only data pointers. Do you have an idea how not to allocate? Except having many queues one-per-each-function?
Template is fine here for me, easier to extend if necessary in the future, but up to you.

Reply:
Now I saw that they everywhere use stubborn_dispatch_uni (where they pass function and pointer, and pointer to data is thread_context, which is not allocated in this function! So they really need xtm_fun_dispatch